packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp

```diff
@@ -1076,47 +1231,279 @@ namespace YAMLParameterList
+      try
+      {
+        quoted_as<int>(first_value);
+        using arr_t = Teuchos::Array<Teuchos::Array<int>>;
+        if (is_ragged) {
+          safe_set_entry<arr_t>(parent, key, getYaml2DRaggedArray<arr_t, int>(node, ndim, key));
+        } else {
+          safe_set_entry<Teuchos::TwoDArray<int>>(parent, key, getYamlTwoDArray<int>(node));
+        }
+      }
+      catch(...)
+      {
+        try
+        {
+          quoted_as<double>(first_value);
+          using arr_t = Teuchos::Array<Teuchos::Array<double>>;
+          if (is_ragged) {
+            safe_set_entry<arr_t>(parent, key, getYaml2DRaggedArray<arr_t, double>(node, ndim, key));
+          } else {
+            safe_set_entry<Teuchos::TwoDArray<double>>(parent, key, getYamlTwoDArray<double>(node));
+          }
+        }
+        catch(...)
+        {
+          try
+          {
+            quoted_as<std::string>(first_value);
+            using arr_t = Teuchos::Array<Teuchos::Array<std::string>>;
+            if (is_ragged) {
+              safe_set_entry<arr_t>(parent, key, getYaml2DRaggedArray<arr_t, std::string>(node, ndim, key));
+            } else {
+              safe_set_entry<Teuchos::TwoDArray<std::string>>(parent, key, getYamlTwoDArray<std::string>(node));
+            }
+          }
+          catch(...)
+          {
+            throw YamlSequenceError(std::string("TwoDArray \"") + key + "\" must contain int, double, bool or string");
+          }
+        }
+      }
+    }
+    else if (ndim == 3)
+    {
+      ::YAML::Node const& first_value = *(node.begin()->begin()->begin());
+      try
+      {
+        quoted_as<int>(first_value);
+        using arr_t = Teuchos::Array<Teuchos::Array<Teuchos::Array<int>>>;
+        safe_set_entry<arr_t>(parent, key, getYaml3DArray<arr_t, int>(node, ndim, key));
+      }
+      catch(...)
+      {
+        try
+        {
+          quoted_as<double>(first_value);
+          using arr_t = Teuchos::Array<Teuchos::Array<Teuchos::Array<double>>>;
+          safe_set_entry<arr_t>(parent, key, getYaml3DArray<arr_t, double>(node, ndim, key));
+
+        }
+        catch(...)
+        {
+          try
+          {
+            quoted_as<std::string>(first_value);
+            using arr_t = Teuchos::Array<Teuchos::Array<Teuchos::Array<std::string>>>;
+            safe_set_entry<arr_t>(parent, key, getYaml3DArray<arr_t, std::string>(node, ndim, key));
+
+          }
+          catch(...)
+          {
+            throw YamlSequenceError(std::string("3DArray \"") + key + "\" must contain int, double, bool or string");
+          }
+        }
+      }
+    }
+  }
+  else if(node.Type() == ::YAML::NodeType::Null)
+  {
+    // treat NULL as empty string (not an error)
+    safe_set_entry<std::string>(parent, key, std::string());
+  }
+  else
+  {
+    // Undefined
+    throw YamlUndefinedNodeError("Value type in a key-value pair must be one of: int, double, string, array, sublist.");
```

Comment:
I could not think of a way to test this exception--I have left a comment in the unit test